nixos/modules/tasks/filesystems/zfs.nix

```diff
@@ -633,23 +633,39 @@ in
         scriptArgs = "%i";
         path = [ pkgs.gawk cfgZfs.package ];
 
-        # Note this script attempts to resize more than just devices.
-        #
-        # It will also try to `online -e` the pool itself and every
-        # vdev. This will print otherwise harmless errors.
-        #
-        # ZFS has no way of enumerating _just_ devices in a pool in a way
-        # that `zpool online -e` supports.
-        #
+        # ZFS has no way of enumerating just devices in a pool in a way
+        # that 'zpool online -e' supports. Thus, we've implemented a
+        # bit of a strange approach of highlighting just devices.
         # See: https://github.com/openzfs/zfs/issues/12505
-        script = ''
+        script = let
+          zpoolScripts = pkgs.writeScriptBin "device-highlighter" ''
+            #!/bin/sh
+
+            # This UUID has been chosen at random and is to provide a
+            # collision-proof, predictable token to search for
+            echo "THIS-IS-A-DEVICE-37108bec-aff6-4b58-9e5e-53c7c9766f05"
+          '';
+        in ''
           pool=$1
 
-          echo "Expanding all devices for $pool"
+          echo "Expanding all devices for $pool."
+
+          # Put our device-highlighter script it to the PATH
+          export ZPOOL_SCRIPTS_PATH=${zpoolScripts}/bin
 
-          for device in $(zpool list -vH "$pool" | awk '($10 = "ONLINE") {print $1;}'); do
-            echo "Attempting to expand $device of $pool"
-            zpool online -e "$pool" "$device" || true
+          # Enable running our precisely specified zpool script as root
+          export ZPOOL_SCRIPTS_AS_ROOT=1
```

Comment:
Since I was a bit skeptical of this, I looked up why it's necessary:

```
Allow a privileged user to run zpool
status/iostat -c.  Normally, only unprivileged
users are allowed to run -c.
```

(basically, we wouldn't be able to run the "marker" script without this)